plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+// Package openstack implements an OpenStack input plugin for Telegraf
+//
+// The OpenStack input plug is a simple two phase metric collector.  In the first
+// pass a set of gatherers are run against the API to cache collections of resources.
+// In the second phase the gathered resources are combined and emitted as metrics.
+//
+// No aggregation is performed by the input plugin, instead queries to InfluxDB should
+// be used to gather global totals of things such as tag frequency.
+package openstack
+
+import (
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/hex"
+	"fmt"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/extensions/schedulerstats"
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/extensions/volumetenants"
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v2/volumes"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/aggregates"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/diagnostics"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/hypervisors"
+	nova_services "github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/services"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/flavors"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/projects"
+	"github.com/gophercloud/gophercloud/openstack/identity/v3/services"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/agents"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/subnets"
+	"github.com/gophercloud/gophercloud/openstack/orchestration/v1/stacks"
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"log"
+	"net/http"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+const (
+	// plugin is used to identify ourselves in log output
+	plugin = "openstack"
+)
+
+// serviceType is an OpenStack service type
+type serviceType string
+
+const (
+	volumeV2Service serviceType = "volumev2"
+	Orchestration   serviceType = "orchestration"
+)
+
+// tagMap maps a tag name to value.
+type tagMap map[string]string
+
+// fieldMap maps a field to an arbitrary data type.
+type fieldMap map[string]interface{}
+
+// serverDiag is a map of server diagnostics fields
+type serverDiag map[string]interface{}
+
+// serviceMap maps service id to a Service struct.
+type serviceMap map[string]services.Service
+
+// stackMap maps stack id to a Stack struct.
+type stackMap map[string]stacks.ListedStack
+
+// nova_serviceMap maps nova_service id to a nova Service struct.
+type nova_serviceMap map[string]nova_services.Service
```

Comment:
Please use those directly in the code instead of declaring them here. This way the type will be obvious in the code.